feat(utxo-lib): refactor psbt input updates out of add

davidkaplanbitgo opened the pull request:
By refactoring out the input updates out from adding the inputs to the PSBT, we are able to add missing information to a PSBT if the one we are given is incomplete.

In addition, there is a fix to adding Musig2 nonces so that they are at the right input.

BG-72538

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

saravanan7mani commented on modules/utxo-lib/src/bitgo/wallet/Unspent.ts:
```diff
@@ -246,3 +302,18 @@ export function addWalletUnspentToPsbt(
+  network: Network = psbt.network
+): void {
+  addBaseUnspentToPsbt(psbt, u, network);
+  updateWalletUnspentForPsbt(psbt, psbt.inputCount - 1, u, rootWalletKeys, signer, cosigner);
```
Does PSBT object allows updating if data already exist?

saravanan7mani commented on modules/utxo-lib/src/bitgo/wallet/Unspent.ts:
```diff
@@ -120,57 +120,113 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
+  inputIndex: number,
+  u: Unspent<bigint> | WalletUnspent<bigint>
+): boolean {
+  if (inputIndex < 0 || inputIndex >= psbt.inputCount) {
```
`checkForInput(psbt.data.inputs, inputIndex)` can be used.

OttoAllmendinger commented on modules/utxo-lib/src/bitgo/wallet/Unspent.ts:
```diff
@@ -120,129 +122,220 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
+  }
+}
+
+function addBaseUnspentToPsbt(psbt: UtxoPsbt, u: Unspent<bigint>): void {
```
```suggestion
function addUnspentToPsbt(psbt: UtxoPsbt, u: Unspent<bigint>): void {
```

OttoAllmendinger commented on modules/utxo-lib/src/bitgo/wallet/Unspent.ts:
```diff
@@ -120,129 +122,220 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
+ * @param u
+ * @param redeemScript Only overrides if there is no redeemScript in the input currently
+ */
+export function updateReplayProtectionUnspentToPsbt(
```
```suggestion
export function updatePsbtRedeemScript(
```

OttoAllmendinger reviewed (changes requested):
few naming suggestions

davidkaplanbitgo commented on modules/utxo-lib/src/bitgo/wallet/Unspent.ts:
```diff
@@ -120,129 +122,220 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
 }
 
-export function addWalletUnspentToPsbt(
+export function addReplayProtectionUnspentToPsbt(psbt: UtxoPsbt, u: Unspent<bigint>, redeemScript: Buffer): void {
```
Not sure if it is a good idea to change the name of this function in this PR - would have breaking changes

OttoAllmendinger replied:
good point

davidkaplanbitgo committed:
fix(utxo-lib): correctly get indices for setMusig2NoncesInner

When calling setInputMusig2Nonce/HD, we are only setting the nonces
for a single input, and we need to use the index of the input in the
psbt, not the index of the input in the `inputs` array.

BG-72538

davidkaplanbitgo committed:
feat(utxo-lib): refactor psbt input update from adding

BG-72538

OttoAllmendinger commented on modules/utxo-lib/src/bitgo/wallet/Unspent.ts:
```diff
@@ -120,129 +130,232 @@ export interface WalletUnspentLegacy<TNumber extends number | bigint = number> e
   if (!isUnspentWithPrevTx(u) && !isZcash) {
     throw new Error('Error, require previous tx to add to PSBT');
   }
+  if (isZcash && !input.witnessUtxo) {
+    const { script, value } = toPrevOutput(u, psbt.network);
+    psbt.updateInput(psbt.inputCount - 1, { witnessUtxo: { script, value } });
+  } else if (!isZcash && !input.nonWitnessUtxo) {
+    psbt.updateInput(psbt.inputCount - 1, { nonWitnessUtxo: (u as UnspentWithPrevTx<bigint>).prevTx });
+  }
```
shouldn't you use `inputIndex` here?

OttoAllmendinger commented on modules/utxo-lib/test/bitgo/psbt/Psbt.ts:
```diff
@@ -340,6 +354,107 @@ describe('isPsbt', function () {
+    signAllInputs(psbt);
+  });
+
+  const componentsOnEachScriptType = {
+    p2sh: ['nonWitnessUtxo', 'redeemScript', 'bip32Derivation'],
+    p2shP2wsh: ['witnessUtxo', 'bip32Derivation', 'redeemScript', 'witnessScript'],
+    p2wsh: ['witnessUtxo', 'witnessScript', 'bip32Derivation'],
+    p2tr: ['witnessUtxo', 'tapLeafScript', 'tapBip32Derivation'],
+    p2trMusig2: ['witnessUtxo', 'tapBip32Derivation', 'tapInternalKey', 'tapMerkleRoot', 'unknownKeyVals'],
+    p2shP2pk: ['redeemScript', 'nonWitnessUtxo'],
+  };
+  inputScriptTypes.forEach((scriptType, i) => {
+    componentsOnEachScriptType[scriptType].forEach((inputComponent) => {
```
nice test
